Auto detect and link graph inputs for `graph` tasks

superstar54 opened the pull request:
Resolves #603.

## Considerations for detecting graph-level inputs

### Pros

* Clear visualization of graph inputs, similar to the context manager approach.
* When the WorkGraph runs, graph-level inputs are serialized and passed to tasks, so the data flow within the AiiDA process can be tracked. Otherwise, tasks would serialize their raw inputs, creating multiple distinct AiiDA data nodes for the same data.
* Serialization requires users to declare the input structure, which makes graphs more robust and less error prone.

### Cons

* The serialization requirement forces users to declare the input structure, which raises the learning curve for new users. They also need to understand the concept of namespaces.

## Implementation

There are many ways to track inputs. One option is static analysis with Python’s `ast` module, as mentioned in #603. However, static analysis cannot see branches and values that only materialize at runtime.

In this PR, we record the `id` of input values and build `value` to `id` mappings for the graph-level inputs and for every task’s inputs. Using these mappings, we can detect which tasks each input value flows into. Since this happens at runtime, it captures complex control flow.

Example:

```python
@task
def add(x, y):
    return x + y

@task
def multiply(x, y):
    return x * y

@task.graph
def add_multiply(a, b, c):
    sum = add(x=a, y=b).result
    if a > 0:
        a = add(x=a, y=b).result
        product = multiply(a, y=c)
    else:
        product = multiply(a, y=c)
    return product

wg = add_multiply.build_graph(a=1, b=2, c=3)
wg
```

If `a = -1`, the branch and data flow change accordingly.

## Limitations of the current approach

In Python:

```python
x = 1
y = x
add(1, y=y)
```

Here `x`, `y`, and `1` can share the same identity, so you cannot distinguish variable names from the literal value using identity alone. Using `ast` helps recover variable names from source, but covering all real-world cases is difficult, and correctness is hard to guarantee.

## Alternative approach

* Define `inputs` in the decorator.
* Inside the function, use `wg.inputs` explicitly.

This is opt in. If users do not reference `wg.inputs`, this mode is not activated and `wg.inputs` is cleared.

```python
@task.graph(inputs={"nested": {"identifier": "workgraph.namespace",
                               "sockets": {"a": {"identifier": "any"},
                                           "b": {"identifier": "any"}}}})
def AddMultiply(x, y, nested: dict):
    wg = get_current_graph()
    x = add(x=wg.inputs.x, y=wg.inputs.y).result
    y = add(x=wg.inputs.nested["a"], y=wg.inputs.nested["b"]).result
    return multiply(x=x, y=y).result

wg = AddMultiply.build_graph(
    x=1,
    y=2,
    nested={"a": 1, "b": 3},
)
```

superstar54 committed:
Auto detect and link graph inputs for `graph` tasks

superstar54 committed:
Merge branch 'main' into auto_detect_graph_inputs

superstar54 committed:
Resolve the links from the graph inputs to the inputs of the tasks based on the mapping

superstar54 committed:
remove the inputs socket of "graph_inputs"

edan-bainglass reviewed (approved):
LGTM! Once the tests pass... 😅

superstar54 committed:
Bump node-graph to 0.2.20

superstar54 committed:
Update test

superstar54 committed:
handle case multiple graph inputs share the same value

Finds the corresponding graph input socket for a given task socket.

Because we use the id of the input value as the identifier for the mapping, there are two cases to consider:
1. The value is unambiguous (only one graph input has this value).
2. The value is ambiguous (multiple graph inputs have the same value). In this case,
   we uses the source-code variable name to find the correct socket.

superstar54 committed:
Fix eos doc

edan-bainglass reviewed (commented):
Just one comment for now. Need to test.
